extensions/amp-analytics/0.1/cross-domain-iframe-message-queue.js

```diff
@@ -0,0 +1,164 @@
+    this.queue_ = [];
+
+    /** @type {?number} */
+    this.timerId_ = null;
+
+    /** @type {boolean} */
+    this.isReady_ = false;
+
+    /** @type {!../../../3p/iframe-messaging-client.IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @type {string} */
+    this.envelopeType_ = envelopeType;
+  }
+
+  /**
+   * Destructor
+   */
+  destroy() {
+    CrossDomainIframeMessageQueue.timer_.cancel(this.timerId_);
+    this.timerId_ = null;
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   * associated data
+   */
+  setIsReady() {
+    this.isReady_ = true;
+    this.sendQueuedMessages_();
+  }
+
+  /**
+   * Builds an event message to be passed to the cross-domain iframe(s)
+   * @param {!string} senderId
+   * @param {!string} message The data to send
+   * @return {../../../src/3p-analytics-common.ampAnalytics3pEvent}
+   */
+  static buildEventMessage(senderId, message) {
+    const messageObject = {
+      senderId,
+      type: AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+    };
+    messageObject[AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT] = message;
+    const typedMessageObject =
+      /** @type {../../../src/3p-analytics-common.ampAnalytics3pEvent} */
+      (messageObject);
+    return typedMessageObject;
+  }
+
+  /**
+   * Builds a message indicating that a new creative is now using the
+   * cross-domain iframe. The message may include some extra config data.
+   * @param {!string} senderId
+   * @param {string=} opt_extraData The data to send to the frame
+   * @returns {../../../src/3p-analytics-common.ampAnalytics3pNewCreative}
+   */
+  static buildNewCreativeMessage(senderId, opt_extraData) {
+    const messageObject = {
+      senderId,
+      type: AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE,
+    };
+    // Still send the message to indicate there is a new creative
+    opt_extraData = opt_extraData || '';
+    messageObject[AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE] = opt_extraData;
+    const typedMessageObject =
+      /** @type {../../../src/3p-analytics-common.ampAnalytics3pNewCreative} */
+      (messageObject);
+    return typedMessageObject;
+  }
+
+  /**
+   * Enqueues a message (event or extra data) to be sent to a cross-domain
+   * iframe.
+   * @param {!(
+   *           ../../../src/3p-analytics-common.ampAnalytics3pEvent|
+   *           ../../../src/3p-analytics-common.ampAnalytics3pNewCreative)
+   *        } messageObject
+   * The data to be enqueued and then sent to the iframe
+   */
+  enqueue(messageObject) {
+    if (this.queue_.length > MAX_QUEUE_SIZE_) {
+      dev().warn(TAG_, 'Queue has exceeded maximum size');
+      this.queue_.shift();
+    }
+    this.queue_.push(messageObject);
+    if (this.timerId_) {
+      return; // Timer is already about to fire, no need to set a new one
+    }
+    this.timerId_ = CrossDomainIframeMessageQueue.timer_.delay(() => {
```

Comment:
This still suffers from the same problem where message after long inactivity will result in it being unnecessarily delayed.  I recommend moving to a setInterval approach which does nothing if the queue is empty.  Seems simpler to me and gives the correct behavior of throttling.